File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        m=float('-inf')
        self.count=0
        def dfs(root,m):
            if not root:
                return
            if root.val>=m:
                m=root.val
                self.count+=1
            dfs(root.left,m)
            dfs(root.right,m)
        dfs(root,m)
        return self.count

File: Data Structures & Algorithms/balanced-binary-tree/submission-5.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int result=0;
    public boolean isBalanced(TreeNode root) {
        maxDepth(root);
        return result<=1;
    }
    public int maxDepth(TreeNode root){
        if(root == null) return 0;
        int lh = maxDepth(root.left);
        int rh = maxDepth(root.right);
        result = Math.max(Math.abs(lh-rh),result);
        return Math.max(lh,rh)+1;
    }
}
